File: cfms_bd/app/Http/Controllers/Admin/BatchController.php
<?php

namespace App\Http\Controllers\Admin;
use App\Http\Controllers\Controller;
use App\Models\Batch;
use Illuminate\Http\Request;
use Illuminate\Validation\Rule;
use Illuminate\Support\Facades\Validator;

class BatchController extends Controller
{
    /**
     * 1. Get all batches (with their program details)
     */
    public function index()
    {
        // Eager load the program to avoid N+1 query performance issues
        $batches = Batch::with('program')->get();
        
        return response()->json([
            'success' => true,
            'data' => $batches
        ], 200);
    }

    /**
     * 2. Store a newly created batch
     */
    public function store(Request $request)
    {
        // Validation logic handling the composite unique key
        $validator = Validator::make($request->all(), [
            'batch_name' => 'required|string|max:255',
            'program_id' => [
                'required',
                'exists:programs,id',
                Rule::unique('batches')->where(function ($query) use ($request) {
                    return $query->where('batch_name', $request->batch_name)
                                 ->where('program_id', $request->program_id);
                })
            ],
        ], [
            // Custom English error message for duplicate batch in the same program
            'program_id.unique' => 'This batch name already exists for the selected program.'
        ]);

        if ($validator->fails()) {
            return response()->json([
                'success' => false,
                'message' => 'Validation Error',
                'errors' => $validator->errors()
            ], 422);
        }

        try {
            $batch = Batch::create([
                'batch_name' => $request->batch_name,
                'program_id' => $request->program_id,
            ]);

            return response()->json([
                'success' => true,
                'message' => 'Batch created successfully.',
                'data' => $batch
            ], 201);

        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Failed to create batch.',
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * 3. Show a specific batch
     */
    public function show($id)
    {
        $batch = Batch::with('program')->find($id);

        if (!$batch) {
            return response()->json([
                'success' => false,
                'message' => 'Batch not found.'
            ], 404);
        }

        return response()->json([
            'success' => true,
            'data' => $batch
        ], 200);
    }

    /**
     * 4. Update the specified batch
     */
    public function update(Request $request, $id)
    {
        $batch = Batch::find($id);

        if (!$batch) {
            return response()->json([
                'success' => false,
                'message' => 'Batch not found.'
            ], 404);
        }

        // Validation logic for update, ignoring the current batch's ID in the unique check
        $validator = Validator::make($request->all(), [
            'batch_name' => 'sometimes|required|string|max:255',
            'program_id' => [
                'sometimes',
                'required',
                'exists:programs,id',
                Rule::unique('batches')->where(function ($query) use ($request, $batch) {
                    return $query->where('batch_name', $request->batch_name ?? $batch->batch_name)
                                 ->where('program_id', $request->program_id ?? $batch->program_id);
                })->ignore($batch->id)
            ],
        ], [
            'program_id.unique' => 'This batch name already exists for the selected program.'
        ]);

        if ($validator->fails()) {
            return response()->json([
                'success' => false,
                'message' => 'Validation Error',
                'errors' => $validator->errors()
            ], 422);
        }

        try {
            $batch->update($request->all());

            return response()->json([
                'success' => true,
                'message' => 'Batch updated successfully.',
                'data' => $batch
            ], 200);

        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Failed to update batch.',
                'error' => $e->getMessage()
            ], 500);
        }
    }

    /**
     * 5. Remove the specified batch
     */
    public function destroy($id)
    {
        $batch = Batch::find($id);

        if (!$batch) {
            return response()->json([
                'success' => false,
                'message' => 'Batch not found.'
            ], 404);
        }

        try {
            $batch->delete();

            return response()->json([
                'success' => true,
                'message' => 'Batch deleted successfully.'
            ], 200);

        } catch (\Exception $e) {
            return response()->json([
                'success' => false,
                'message' => 'Failed to delete batch.',
                'error' => $e->getMessage()
            ], 500);
        }
    }
}

File: cfms_fd/lib/ViewModel/Controller/AdminController/BatchController.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class BatchController {
  // Base API URL
  final String baseUrl = 'http://127.0.0.1:8000/api';

  // 1. Fetch All Batches
  Future<Map<String, dynamic>> fetchBatches(String token) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/admin/batches'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final decodedData = json.decode(response.body);

        if (decodedData is List) {
          return {'success': true, 'data': decodedData};
        } else if (decodedData is Map<String, dynamic>) {
          return {
            'success': decodedData['success'] ?? true,
            'data': decodedData['data'] ?? decodedData['batches'] ?? [],
          };
        }
        return {'success': true, 'data': decodedData};
      } else {
        return {'success': false, 'message': 'Failed to load batches from database.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Database Error: $e'};
    }
  }

  // 2. Fetch All Programs Dynamically
  Future<Map<String, dynamic>> fetchPrograms(String token) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/admin/program'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final decodedData = json.decode(response.body);

        if (decodedData is List) {
          return {'success': true, 'data': decodedData};
        } else if (decodedData is Map<String, dynamic>) {
          return {
            'success': decodedData['success'] ?? true,
            'data': decodedData['data'] ?? decodedData['programs'] ?? [],
          };
        }
        return {'success': true, 'data': decodedData};
      } else {
        return {'success': false, 'message': 'Failed to load programs from database.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Database Error: $e'};
    }
  }

  // 3. Fetch All Departments Dynamically
  Future<Map<String, dynamic>> fetchDepartments(String token) async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl/admin/department'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
        },
      );

      if (response.statusCode == 200) {
        final decodedData = json.decode(response.body);

        if (decodedData is List) {
          return {'success': true, 'data': decodedData};
        } else if (decodedData is Map<String, dynamic>) {
          return {
            'success': decodedData['success'] ?? true,
            'data': decodedData['data'] ?? decodedData['departments'] ?? decodedData['department'] ?? [],
          };
        }
        return {'success': true, 'data': decodedData};
      } else {
        return {'success': false, 'message': 'Failed to load departments from database.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Database Error: $e'};
    }
  }

  // 4. Create New Batch
  Future<Map<String, dynamic>> createBatch(String token, String batchName, String programId) async {
    try {
      final response = await http.post(
        Uri.parse('$baseUrl/admin/batches'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'batch_name': batchName,
          'program_id': programId,
        }),
      );

      final data = json.decode(response.body);

      if (response.statusCode == 201 || response.statusCode == 200) {
        return {
          'success': data['success'] ?? true,
          'message': data['message'] ?? 'Batch created successfully.',
          'data': data['data']
        };
      } else if (response.statusCode == 422) {
        String errorMsg = data['message'] ?? 'Validation failed.';
        if (data['errors'] != null && data['errors']['program_id'] != null) {
          errorMsg = data['errors']['program_id'][0];
        }
        return {'success': false, 'message': errorMsg};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Failed to create batch.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Error: $e'};
    }
  }

  // 5. Update Existing Batch
  Future<Map<String, dynamic>> updateBatch(String token, int id, String batchName, String programId) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl/admin/batches/$id'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
          'Content-Type': 'application/json',
        },
        body: json.encode({
          'batch_name': batchName,
          'program_id': programId,
        }),
      );

      final data = json.decode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': data['success'] ?? true,
          'message': data['message'] ?? 'Batch updated successfully.',
          'data': data['data']
        };
      } else if (response.statusCode == 422) {
        String errorMsg = data['message'] ?? 'Validation failed.';
        if (data['errors'] != null && data['errors']['program_id'] != null) {
          errorMsg = data['errors']['program_id'][0];
        }
        return {'success': false, 'message': errorMsg};
      } else {
        return {'success': false, 'message': data['message'] ?? 'Failed to update batch.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Error: $e'};
    }
  }

  // 6. Delete Batch
  Future<Map<String, dynamic>> deleteBatch(String token, int id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl/admin/batches/$id'),
        headers: {
          'Authorization': 'Bearer $token',
          'Accept': 'application/json',
        },
      );

      final data = json.decode(response.body);
      if (response.statusCode == 200) {
        if (data is Map<String, dynamic>) {
          return {'success': data['success'] ?? true, 'message': data['message'] ?? 'Deleted successfully.'};
        }
        return {'success': true, 'message': 'Batch deleted successfully.'};
      } else {
        return {'success': false, 'message': 'Failed to delete batch from server.'};
      }
    } catch (e) {
      return {'success': false, 'message': 'API Error: $e'};
    }
  }
}